NUP-2299 Added platform-conditional dependency on pycapnp==0.5.8 using PEP-508

vitaly-krugl opened the pull request:
Fixes #3263

oxtopus commented:
👍 provided https://ci.numenta.com/browse/NUP-PY81/latest passes and whatever travis-ci issues are resolved.

Thanks!

oxtopus commented:
Seems to be a problem on OS X:

```
+ python setup.py bdist_wheel
error in nupic setup command: Invalid environment marker: platform_system=='Linux' or platform_system=='Darwin'
```

vitaly-krugl commented:
@oxtopus : I'll work through this. The problem is that pip and setuptools installed by travis are too old. Travis sets up its python packages in a strange way that makes it difficult to upgrade pip and setuptools, because sys.path lists their eggs explicitly at absolute paths. Even though the upgrade claims that it succeeded, the ones that get used are the old ones!

I think I need to use virtualenv myself to get the environment and package versions that I need to make the build work. This works fine with newer pip/setuptools.

I am also struggling to find meaningful documentation on travis's .travis.yml syntax. They are either  making it difficult to find or it just doesn't exist.

oxtopus commented:
The error I quoted was from the bamboo run at  https://ci.numenta.com/browse/NUP-PY81 in which a recent version of setuptools is properly installed.  If it is a version incompatibility, we should document that and ensure that specific version gets installed.

vitaly-krugl commented on ci/travis/after_success-release-linux.sh:
```diff
@@ -25,11 +25,11 @@ echo "Running after_success-release.sh..."
 echo
 
 echo "Installing boto..."
-pip install boto --user || exit
```
virtualenv activation (see changes in .travis.yml) doesn't allow `--user`

vitaly-krugl commented on ci/travis/before_install-linux.sh:
```diff
@@ -32,21 +32,18 @@ if [ $CC == 'gcc' ]; then
 fi
 
-echo ">>> Installing nupic-linux64..."
-git clone https://github.com/numenta/nupic-linux64.git
```
The nupic-linuxXX and nupic-darwinXX environments have gotten hopelessly out of date, and are not needed for the build to succeed.

CC @scottpurdy

vitaly-krugl commented:
Please review. This addresses the P2 issue NUP-2299 "Bamboo CI pipeline red"

oxtopus commented on ci/bamboo/build-osx.sh:
```diff
@@ -9,6 +9,12 @@ export PYTHONPATH=${HOME}/Library/Python/2.7/lib/python/site-packages:${PYTHONPA
 
+# Upgrade setuptools (for PEP-508 support used in extras_require); also wheel
+pip install --user --upgrade --ignore-installed setuptools wheel
+python -c 'import pip; print "pip version=", pip.__version__'
```
Please remove these lines, or add them to `ci/bamboo/build-linux.sh` for consistency.

oxtopus commented:
To reiterate, if there's a specific version requirement on setuptools, I'd like to see it documented in this changeset.  You put a lot of work into making the following changes to setup.py work consistently between travis and bamboo -- going above and beyond what was required -- and it's likely someone will encounter the same problems, too, without that version or greater.

```diff
 +    extras_require = {
 +      # Default requirement based on system type
 +      ":platform_system=='Linux' or platform_system=='Darwin'":
 +        ["pycapnp==0.5.8"],
 +
 +      # Superseded by platform_system-conditional requirement, but keeping
 +      # empty extra for compatibility as recommended by setuptools doc.
 +      "capnp": [],
 +    },
```

oxtopus commented:
In retrospect, `pip install --upgrade --ignore-installed setuptools wheel` should never have made it into `ci/bamboo/build-linux.sh` (I put it there.  I'm sorry).  In the bamboo builds, `get-pip.py` is used to install `pip`, `setuptools`, and `wheel` -- see https://pip.pypa.io/en/stable/installing/#installing-with-get-pip-py for official recommended installation technique.  We're vendoring in `get-pip.py` (see https://github.com/numenta/nupic/tree/master/ci/bamboo) for a number of reasons, one of which is to version lock a known, good version of `pip`, `setuptools`, and `wheel`.  Calling `pip install --upgrade --ignore-installed setuptools wheel` defeats the purpose of bundling `get-pip.py`.

Instead, if we need to update `pip`, `setuptools`, or `wheel`, we should update `get-pip.py` (and document the versions for anyone not using the correct installation method).

vitaly-krugl commented:
@oxtopus, where is the current location that documents the versions of pip, setuptools, and wheel?

oxtopus commented:
https://github.com/numenta/nupic/blob/master/README.md#dependencies, https://github.com/numenta/nupic/blob/master/external/common/requirements.txt, and https://github.com/numenta/nupic/blob/master/DEPENDENCIES.md are all means of conveying dependencies in nupic, none of which currently specify any specific version of setuptools.  Since there is now a dependency on a specific version of setuptools, we need to make that clear in the documentation as as been done for other dependencies where it's known that a specific version (or later) is required.

Personally, modifying `setup.py` to assert on a specific or later version, or a try/except block to catch the failure and suggest that the user install the specific version or later of setuptools would be most effective, but is not necessary in this case -- as long as it's documented.

vitaly-krugl commented on ci/bamboo/install-pip-setuptools-wheel.sh:
```diff
@@ -0,0 +1,49 @@
+#!/bin/bash
```
Shared script for installing specific versions of pip, setuptools, and wheel on *nix (Linux and OS X) builds.

vitaly-krugl commented:
@oxtopus, regarding
> Instead, if we need to update pip, setuptools, or wheel, we should update `get-pip.py`

We don't want to modify `get-pip.py`. Instead, since `get-pip.py` accepts all command-line options that you can pass to `pip`, we simply pass the required packages (pip, setuptools, wheel) along with their versions to `get-pip.py` using a script that is shared between the builds.

vitaly-krugl commented:
This is ready for code review once again. I removed upgrading of pip/setuptools/wheel, installed explicit version numbers of those packages, and updated version requirements in the README.

vitaly-krugl commented:
Incidentally, the goal for the chosen approach is to isolate the logic of deciding **whether** and **which version** of pycapnp to install in a single location, thus minimizing the likelihood of bugs and simplifying maintenance.

```
 +    extras_require = {
 +      # Default requirement based on system type
 +      ":platform_system=='Linux' or platform_system=='Darwin'":
 +        ["pycapnp==0.5.8"],
 +
 +      # Superseded by platform_system-conditional requirement, but keeping
 +      # empty extra for compatibility as recommended by setuptools doc.
 +      "capnp": [],
 +    },
```

vitaly-krugl commented:
PR builds were successful at 067978623f0da3525bea06ec11d2b3ea6ba616a1 on Travis, AppVeyor, and Bamboo (https://ci.numenta.com/browse/NUP-PY81-20/log)

oxtopus commented:
Thanks!  I appreciate the effort you went through here.  `install-pip-setuptools-wheel.sh` adds an unnecessary amount of complexity, but that's a discussion we can take offline and it's not so important that it's worth holding back a solution that would otherwise fix a broken pipeline. 👍 :shipit:

vitaly-krugl commented:
@oxtopus, I would not wish to merge a change that, according to you,
> adds an unnecessary amount of complexity

What do you recommend in its place that would not add an "unnecessary amount of complexity"?

oxtopus commented:
I'd recommended earlier (here, or elsewhere) that all you need to do is add `[capnp]` to the `pip install` command.  For example, the line https://github.com/vitaly-krugl/nupic/blob/a7d3c724911332c8702ed7daaa1f768709faed93/ci/bamboo/build-osx.sh#L17 becomes:

`pip install --user --ignore-installed -f wheelhouse/ dist/nupic-`cat VERSION`*.whl[capnp]`

Anecdotal from my mac:

```
Austins-MacBook-Pro-6:Downloads amarshall$ pip install --user nupic-0.5.4-py2-none-any.whl[capnp]
Processing ./nupic-0.5.4-py2-none-any.whl
Requirement already satisfied (use --upgrade to upgrade): unittest2==0.5.1 in /Users/amarshall/Library/Python/2.7/site-packages (from nupic==0.5.4)
Requirement already satisfied (use --upgrade to upgrade): asteval==0.9.1 in /Users/amarshall/Library/Python/2.7/site-packages (from nupic==0.5.4)
Requirement already satisfied (use --upgrade to upgrade): pytest-cov==1.6 in /Users/amarshall/Library/Python/2.7/site-packages (from nupic==0.5.4)
Requirement already satisfied (use --upgrade to upgrade): validictory==0.9.1 in /Users/amarshall/Library/Python/2.7/site-packages (from nupic==0.5.4)
Requirement already satisfied (use --upgrade to upgrade): numpy==1.9.2 in /Users/amarshall/Library/Python/2.7/lib/python/site-packages (from nupic==0.5.4)
Collecting coverage==3.7.1 (from nupic==0.5.4)
Requirement already satisfied (use --upgrade to upgrade): DBUtils==1.1 in /Users/amarshall/Library/Python/2.7/site-packages (from nupic==0.5.4)
Requirement already satisfied (use --upgrade to upgrade): PyYAML==3.10 in /Users/amarshall/Library/Python/2.7/site-packages (from nupic==0.5.4)
Requirement already satisfied (use --upgrade to upgrade): mock==1.0.1 in /Users/amarshall/Library/Python/2.7/site-packages (from nupic==0.5.4)
Collecting nupic.bindings==0.4.4 (from nupic==0.5.4)
  Using cached nupic.bindings-0.4.4-cp27-none-macosx_10_11_intel.whl
Requirement already satisfied (use --upgrade to upgrade): ordereddict==1.1 in /Users/amarshall/Library/Python/2.7/site-packages (from nupic==0.5.4)
Requirement already satisfied (use --upgrade to upgrade): PyMySQL==0.6.2 in /Users/amarshall/Library/Python/2.7/site-packages (from nupic==0.5.4)
Requirement already satisfied (use --upgrade to upgrade): pyproj==1.9.3 in /Users/amarshall/Library/Python/2.7/site-packages (from nupic==0.5.4)
Requirement already satisfied (use --upgrade to upgrade): python-dateutil==2.1 in /Users/amarshall/Library/Python/2.7/lib/python/site-packages/python_dateutil-2.1-py2.7.egg (from nupic==0.5.4)
Requirement already satisfied (use --upgrade to upgrade): pytest==2.5.1 in /Users/amarshall/Library/Python/2.7/lib/python/site-packages (from nupic==0.5.4)
Requirement already satisfied (use --upgrade to upgrade): psutil==1.0.1 in /Users/amarshall/Library/Python/2.7/site-packages (from nupic==0.5.4)
Requirement already satisfied (use --upgrade to upgrade): pytest-xdist==1.8 in /Users/amarshall/Library/Python/2.7/site-packages (from nupic==0.5.4)
Requirement already satisfied (use --upgrade to upgrade): prettytable==0.7.2 in /Users/amarshall/Library/Python/2.7/site-packages (from nupic==0.5.4)
Collecting pycapnp==0.5.5; extra == "capnp" (from nupic==0.5.4)
Requirement already satisfied (use --upgrade to upgrade): cov-core>=1.6 in /Users/amarshall/Library/Python/2.7/lib/python/site-packages (from pytest-cov==1.6->nupic==0.5.4)
Requirement already satisfied (use --upgrade to upgrade): six in /Users/amarshall/Library/Python/2.7/site-packages (from python-dateutil==2.1->nupic==0.5.4)
Requirement already satisfied (use --upgrade to upgrade): py>=1.4.19 in /Users/amarshall/Library/Python/2.7/site-packages (from pytest==2.5.1->nupic==0.5.4)
Requirement already satisfied (use --upgrade to upgrade): execnet>=1.0.8 in /Users/amarshall/Library/Python/2.7/site-packages (from pytest-xdist==1.8->nupic==0.5.4)
Installing collected packages: coverage, nupic.bindings, pycapnp, nupic
Successfully installed coverage-4.0a1 nupic-0.5.5.dev0 nupic.bindings-0.4.5.dev0 pycapnp-0.5.5
```

Note the `Collecting pycapnp==0.5.5; extra == "capnp" (from nupic==0.5.4)` and `Installing collected packages: coverage, nupic.bindings, pycapnp, nupic
Successfully installed coverage-4.0a1 nupic-0.5.5.dev0 nupic.bindings-0.4.5.dev0 pycapnp-0.5.5` from the output.

I do like the approach, but I'm somewhat conflicted about it because A) it requires a new version of `setuptools`, and B) `get-pip.py` is being used in a way I hadn't anticipated to add support for the newer `setuptools`.

scottpurdy commented:
By the way, we do need to update setuptools to a fairly recent version. I'm not sure the exact version but some dependencies require something fairly recent (and aren't necessarily able to update it themselves beforehand due to the nature of it):

https://github.com/jparyani/pycapnp#common-problems

And I think we should consider alternatives before we have copies of `get-pip.py` sprinkled throughout all of our repos.

vitaly-krugl commented:
@oxtopus, the problem with relying on something like `nupic-0.5.4-py2-none-any.whl[capnp]` that you recommended is that it spreads nupic's dependency information outside of the nupic package. When relying on the `...[capnp]` feature selection, all of our build scripts (as well as 3rd party users) would then need to know something unnecessary about the dependencies of the nupic package and take explicit action based on it. So, all of our ci/bamboo nupic build scripts would need to know when to pass it and when not to, as would the build scripts in nupic.regression(1) and other build scripts and users.

(1)Note that nupic.regression build presently uses the old way, but soon won't have to.

Regarding your feedback
> get-pip.py is being used in a way I hadn't anticipated to add support for the newer setuptools.

I didn't know about that capability either until I read the documentation and found that it's fully supported. So, it made perfect sense to use it. In https://pip.pypa.io/en/stable/installing/#installing-with-get-pip-py:
> Additionally, get-pip.py supports using the [pip install options](https://pip.pypa.io/en/stable/reference/pip_install/#pip-install-options) and the [general options](https://pip.pypa.io/en/stable/reference/pip/#general-options).

oxtopus commented:
I disagree with the conclusions you've made here.  The functionality that allows for specifying `[capnp]` in the pip command has been there for a long time.  I merely suggested that using the status quo functionality is better than duplicating a version number in a build script.

To @scottpurdy's point, there is some duplication between nupic.core and nupic and this changeset further diverges the two projects with respect to how the `pycapnp` dependency is handled.

re: `get-pip.py`, I suspect it has a short-lived future.  I know @scottpurdy wants it gone, so having a disagreement about it certainly doesn't help, but here goes the rationale: `get-pip.py` was included because it is the recommended way of installing `pip` according to official documentation, and it works consistently across platforms without having to keep track of platform-specific installation techniques.  Bundling it in the repo was done to avoid intermittent failures or security issues related to downloading it every time we run an automated build.

Also, just because something is in the documentation, doesn't mean it makes sense in context.  If we're going to download specific versions of pip, setuptools, and wheel in every build, that certainly adds support to the argument for its removal (i.e., it's redundant).

vitaly-krugl commented:
Other than this, I am going to refrain from adding further comments in this now-merged PR.

The way that nupic.core manages the pycapnp dependency is not ideal. I consider this nupic PR a step in the right direction. There is no need to be concerned about diverging from nupic.core's legacy ways to arrive at something simpler/better. nupic.core could also benefit from a similar simplification.

By the way, this is what the pycapnp picture looks like right now in nupic.core, with just about every build script burdened with intimate knowledge about the specific dependency on `pycapnp==0.5.8`. Also, 3rd party and in-house developers need to be aware of the need to install it and of the correct version to use in order to build nupic.core, and all of these things need to be in sync.
```
~/nta/nupic.core (master)$ git grep 'pycapnp=='
Dockerfile:        pycapnp==0.5.8 \
README.md:    pip install pycapnp==0.5.8
bindings/py/setup.py:    extras_require = {"capnp": ["pycapnp==0.5.8"]},
ci/bamboo/build-linux-debug.sh:    pycapnp==0.5.8 \
ci/bamboo/build-linux-release.sh:    pycapnp==0.5.8 \
ci/bamboo/build-manylinux-prototype.sh:${PYBIN}/pip install pycapnp==0.5.8
ci/bamboo/build-osx.sh:    pycapnp==0.5.8 \
ci/travis/before_install-linux.sh:pip install --user pycapnp==0.5.8 || exit
ci/travis/before_install-osx.sh:pip install pycapnp==0.5.8 --user || exit
```

Compare this to the current nupic:
```
~/nta/nupic (master)$ git grep 'pycapnp=='
Dockerfile:        pycapnp==0.5.8 \
setup.py:        ["pycapnp==0.5.8"],
```
... and the one in Dockerfile is technically no longer necessary, except as a build performance speed-up as it avoids the build time of pycapnp.

rhyolight commented:
@vitaly-krugl @oxtopus I've created a topic in the forum committer lounge for you to continue this discussion (if you wish).
